json_serializable/lib/src/utils.dart

```diff
@@ -15,43 +15,61 @@ import 'package:source_gen/source_gen.dart';
 /// Dart code.
 // TODO: still need handle triple singe/double quotes within `value`
 String escapeDartString(String value) {
-  value = value.replaceAll('\\', r'\\');
+  var hasSingleQuote = false;
+  var hasDoubleQuote = false;
+  var hasDollar = false;
+  var canBeRaw = true;
+
   value = value.replaceAllMapped(_escapeRegExp, (match) {
+    var value = match[0];
+    if (value == "'") {
+      hasSingleQuote = true;
+      return value;
+    } else if (value == '"') {
+      hasDoubleQuote = true;
+      return value;
+    } else if (value == r'$') {
+      hasDollar = true;
+      return value;
+    }
+
+    canBeRaw = false;
     var mapped = _escapeMap[match[0]];
     if (mapped != null) return mapped;
     return _getHexLiteral(match[0]);
   });
 
-  var containsDollar = value.contains(r'$');
-
-  if (value.contains("'")) {
-    if (value.contains('"')) {
-      // `value` contains both single and double quotes.
-      // The only safe way to wrap the content is to escape all of the
-      // problematic characters.
-      var string = value
-          .replaceAll(r'$', r'\$')
-          .replaceAll('"', r'\"')
-          .replaceAll("'", r"\'");
-      return "'$string'";
-    } else if (containsDollar) {
-      // `value` contains "'" and "$", but not '"'.
-      // Safely wrap it in a raw string within double-quotes.
-      return 'r"$value"';
+  if (!hasDollar) {
+    if (hasSingleQuote) {
+      if (!hasDoubleQuote) {
+        return '"$value"';
+      }
+      // something
+    } else {
+      // trivial!
+      return "'$value'";
     }
-    return '"$value"';
-  } else if (containsDollar) {
-    // `value` contains "$", but no "'"
-    // wrap it in a raw string using single quotes
-    return "r'$value'";
   }
 
-  // `value` contains no problematic characters - except for '"' maybe.
-  // Wrap it in standard single-quotes.
-  return "'$value'";
+  if (hasDollar && canBeRaw) {
+    if (hasSingleQuote) {
+      if (!hasDoubleQuote) {
+        // quote it with single quotes!
+        return 'r"$value"';
+      }
+    } else {
+      // quote it with single quotes!
+      return "r'$value'";
+    }
+  }
+
+  // The only safe way to wrap the content is to escape all of the
```

Comment:
I take offense with the word "only" :)
You could potentially use tripple-quotes in some cases (you just don't have the information here to know when).